### ISSUE TITLE
JSON-LD

### ISSUE THREAD

**Issue by javierluraschi** [rstudio/pins#177]
Consider supporting JSON-LD instead of data.txt format.

**Comment by javierluraschi** [comment]
*TL;DR;* we still want to use `data.txt` for the top-level data catalog, mostly to list paths to resources; however, for metadata, like describing columns, etc. we should try to integrate `JSON-LD`.

## Investigation

There is a lot of value in JSON-LD; even for simple things, like a [FAQ datasets](https://developers.google.com/search/docs/data-types/faqpage), I wish we could do something in R like (and maybe is possible already):

```r
faq_data <- list(list(q = "What year is it?", a = "2020"))
schemaorg::write(faq_data, schema = "faqpage", "myfaq.json")
```

Would love to extend that to pins to store the JSON-LD in a particular location, say S3 as follows:

```r
schemaorg::create(faq_data, schema = "faqpage") %>%
  pins::pin(name = "myfaq", board = "github")
```

So far so good and definitely excited to get something like this integrated into pins, at some point.

The trickier piece is this concept of [DataCatalog](https://schema.org/DataCatalog) which Google seems to [suggest we use](https://developers.google.com/search/docs/data-types/dataset#approach).

Apparently, a DataCatalog has also a JSON-LD schema, but things get complicated really fast. First is hard to find examples, I did find one:

[http://dcat.be/](http://dcat.be/)

But then, from [schema.org Wikipedia](https://en.wikipedia.org/wiki/Schema.org) page, looks like there are three formats one can use: Microdata, RDF, or JSON-LD and recommends adding a bunch of additional fields, so for instance, a data catalog for iris and mtcars in data.txt would look like:

```yaml
- name: iris
  path: iris.csv
- name: mtcars
  path: mtcars.csv
```

While, I think, in schema.org would look like:

```html
<script type="application/ld+json">
{ 
  "@context": "http://schema.org/",
  "@type": "DataCatalog",
  "name": "Catalog",
  "datasets": 
  [
    { 
       "@type": "Dataset",
       "value": "https://domainname/iris.csv"
    },
    { 
       "@type": "Dataset",
       "value": "https://domainname/mtcars.csv"
    }
  ]
}
</script>
```

I think? The code above might honestly be wrong and in general, makes me think that DataCatalog is an overly complicated approach to list dataset. It also does not seem to be used that often for discoverability since the JSON-LD is usually embedded into a webpage in a <script> tag that web crawlers can spot, web crawling was the main motivation for schema.org, apparently.

Anyways, now, I do think that data.txt should not attempt to prescribe how to describe datasets. Instead, pins should embrace JSON-LD and schema.org whenever possible. At the very least, by properly integrating with the "official" schema.org R package.

In which case, creating a pin for iris would generate the following data.txt entry:

```yaml
- name: iris
  path: iris.csv
  type: table
  schema: iris.json
```

With iris.json representing a proper schema.org Dataset object.

**Comment by javierluraschi** [comment]
Much has been discussed here... I think `data.txt` is the right call for the index, it needs to be kept simple like other formats like `robots.txt` and `humans.txt`... However, for the actual metadata of each pin, we should get the metadata to be in `JSON-LD` format. This is also confusing since the data set index is named `data.txt` and the file describing each pin is also `data.txt`, just FYI.

However, I don't think this will get addressed in the short term, so closing for now.
